sdk/src/main/scala/com/horizen/account/state/WithdrawalMsgProcessor.scala

```diff
@@ -0,0 +1,195 @@
+import com.horizen.account.proposition.AddressProposition
+
+import com.horizen.account.utils.ZenWeiConverter
+import com.horizen.proposition.MCPublicKeyHashProposition
+
+import com.horizen.utils.{BytesUtils, ListSerializer, ZenCoinsUtils}
+import scorex.crypto.hash.Keccak256
+
+import scala.collection.JavaConverters.seqAsJavaListConverter
+import scala.util.{Failure, Success, Try}
+
+trait WithdrawalRequestProvider {
+  private[horizen] def getListOfWithdrawalReqRecords(epochNum: Int, view: AccountStateView): Seq[WithdrawalRequest]
+}
+
+object WithdrawalMsgProcessor extends AbstractFakeSmartContractMsgProcessor with WithdrawalRequestProvider {
+
+  override val fakeSmartContractAddress: AddressProposition = new AddressProposition(BytesUtils.fromHexString("0000000000000000000011111111111111111111"))
+
+  override def fakeSmartContractCodeHash: Array[Byte] =
+    Keccak256.hash("WithdrawalRequestSmartContractCodeHash")
+
+  val getListOfWithdrawalReqsCmdSig: String = "00"
+  val addNewWithdrawalReqCmdSig: String = "01"
+
+  //TODO Define a proper amount of gas spent for each operation
+  val gasSpentForGetListOfWithdrawalReqsCmd: java.math.BigInteger = java.math.BigInteger.ONE
+  val gasSpentForGetListOfWithdrawalReqsFailure: java.math.BigInteger = java.math.BigInteger.ONE
+  val gasSpentForAddNewWithdrawalReqCmd: java.math.BigInteger = java.math.BigInteger.ONE
+  val gasSpentForAddNewWithdrawalReqFailure: java.math.BigInteger = java.math.BigInteger.ONE
+  val gasSpentForGenericFailure: java.math.BigInteger = java.math.BigInteger.ONE
+
+  val MAX_WITHDRAWAL_REQS_NUM_PER_EPOCH = 3900
+  val DUST_THRESHOLD_IN_WEI: java.math.BigInteger = ZenWeiConverter.convertZenniesToWei(ZenCoinsUtils.getMinDustThreshold(ZenCoinsUtils.MC_DEFAULT_FEE_RATE))
+
+
+  override def canProcess(msg: Message, view: AccountStateView): Boolean = {
+    fakeSmartContractAddress.equals(msg.getTo)
+  }
+
+  override def process(msg: Message, view: AccountStateView): ExecutionResult = {
+    //TODO: check errors in Ethereum, maybe for some kind of errors there a predefined types or codes
+
+    try {
+      if (!canProcess(msg, view)) {
+        new InvalidMessage(new IllegalArgumentException(s"Cannot process message $msg"))
+      }
+      else {
+
+        val functionSig = BytesUtils.toHexString(getOpCodeFromData(msg.getData))
+
+        if (getListOfWithdrawalReqsCmdSig.equals(functionSig)) {
+          getListOfWithdrawalReqRecords(msg, view) match {
+            case Success(result) =>
+              result
+            //TODO log
+            case Failure(exception: IllegalArgumentException) =>
+              new ExecutionFailed(gasSpentForGetListOfWithdrawalReqsFailure, exception)
+            case Failure(exception) =>
+              log.error(s"Error while requesting Withdrawal Request list: ${exception.getMessage}", exception)
+              new ExecutionFailed(gasSpentForGetListOfWithdrawalReqsFailure, new Exception(exception))
```

Comment:
It is not clear to me why we have this separation between exceptions and what is the case for `IllegalArgumentException` ?
Let's discuss